File: app/src/main/java/com/jzoft/ygohelper/biz/impl/ProxyCardLocatorLinked.java
package com.jzoft.ygohelper.biz.impl;

import com.jzoft.ygohelper.biz.ProxyCard;
import com.jzoft.ygohelper.biz.ProxyCardLocator;
import com.jzoft.ygohelper.utils.HttpCaller;

/**
 * Created by jjimenez on 13/10/16.
 */
public class ProxyCardLocatorLinked implements ProxyCardLocator {

    private ProxyCardLocator locator;
    private ProxyCardLocator next;

    private ProxyCardLocatorLinked(ProxyCardLocator locator, ProxyCardLocator next) {
        this.locator = locator;
        this.next = next;
    }

    @Override
    public ProxyCard locate(String location) throws HttpCaller.NotFound {
        if (!hasToLocate(location) && next != null)
            return next.locate(location);
        if (next == null)
            return locator.locate(location);
        else
            return next.locate(locator.locate(location).getUrl());
    }

    @Override
    public boolean hasToLocate(String location) {
        return locator.hasToLocate(location);
    }

    public static ProxyCardLocator buildPatchLocatorLinked(ProxyCardLocator... locators) {
        if (locators.length == 0)
            return null;
        return new ProxyCardLocatorLinked(locators[0], buildPatchLocatorLinked(removeFirst(locators)));
    }

    private static ProxyCardLocator[] removeFirst(ProxyCardLocator[] locators) {
        ProxyCardLocator[] withoutFirst = new ProxyCardLocator[locators.length - 1];
        for (int i = 1; i < locators.length; i++) {
            withoutFirst[i - 1] = locators[i];
        }
        return withoutFirst;
    }
}

File: app/src/main/java/com/jzoft/ygohelper/biz/impl/ProxyCardLocatorLinked.kt
package com.jzoft.ygohelper.biz.impl

import com.jzoft.ygohelper.biz.ProxyCard
import com.jzoft.ygohelper.biz.ProxyCardLocator
import com.jzoft.ygohelper.utils.Caller
import rx.Observable

/**
 * Created by jjimenez on 13/10/16.
 */
class ProxyCardLocatorLinked private constructor(private val locator: ProxyCardLocator, private val next: ProxyCardLocator?) : ProxyCardLocator {

    @Throws(Caller.NotFound::class)
    override fun locate(location: String): Observable<ProxyCard> {
        return if (!hasToLocate(location) && next != null) next.locate(location)
        else locator.locate(location).flatMap {
            if (next != null) next.locate(it.url!!)
            else Observable.just(it)
        }
    }

    override fun hasToLocate(location: String): Boolean {
        return locator.hasToLocate(location)
    }

    companion object {

        fun buildPatchLocatorLinked(vararg locators: ProxyCardLocator): ProxyCardLocator? {
            return if (locators.isEmpty()) null else ProxyCardLocatorLinked(locators[0], buildPatchLocatorLinked(*removeFirst(locators)))
        }

        private fun removeFirst(locators: Array<out ProxyCardLocator>): Array<ProxyCardLocator> {
            val withoutFirst = mutableListOf<ProxyCardLocator>()
            for (i in 1 until locators.size) {
                withoutFirst.add(locators[i])
            }
            return withoutFirst.toTypedArray()
        }
    }
}
